(maint) improve code quality rake tasks

er0ck opened the pull request:
i thought we could avoid the bundle install on the host (previous PR).  but we now use a bunch of gem's provided rake tasks in the rakefile.  so we need the gem or the bundle.  installing the gems in the right version against the host ruby version is a PITA.  
we need to unit test against a couple ruby versions anyway, we've had issues there.  so i'll switch back to rvm.  i'll try to narrow it to versions where binaries are available.  i don't really know what platform the host vm is on.
the bundle on the host tends to pollute the bundle in the container, but i believe i've largely tackled that.

more to come...

Commit:
(maint) update to newer rubocop rules

* mostly the defaults
* copied the rest from doctor_teeth

Commit:
(maint) improve code quality rake tasks

* apply uniformity with doctor_teeth
* support everything else, remove backwards compat with name of test tasks

Commit:
(maint) git ignore yardstick reports

Commit:
(maint) turn on coverage, add coveralls functionality

samwoods1 commented on Rakefile:
```diff
@@ -1,29 +1,122 @@
+  desc "Clean/remove the generated YARD Documentation cache"
+  task :clean do
+    original_dir = Dir.pwd
+    Dir.chdir(File.expand_path(File.dirname(__FILE__)))
```
Why are we changing directories? We can remove a directory without cd'ing into it.

er0ck replied:
randell asked the same thing.  i need to comment this, because i keep forgetting myself.  i think the idea is that rake can be run from any dir below project root, so we never know what dir we're in.  but we know _where_ the rakefile is with `__FILE__`.  but i guess we don't have to cd to the dir.  i can just absolute path the yarddir with the __FILE__ path.

samwoods1 commented on Rakefile:
```diff
@@ -1,29 +1,122 @@
+  task :clean do
+    original_dir = Dir.pwd
+    Dir.chdir(File.expand_path(File.dirname(__FILE__)))
+    sh "rm -rf #{YARD_DIR}"
```
Why use shell? We should use FileUtils.rm_rf

samwoods1 commented on Rakefile:
```diff
@@ -1,29 +1,122 @@
+  #   But we still have to send the output to the graphviz processor, etc.
+  task arch: [:yard] do
+    original_dir = Dir.pwd
+    Dir.chdir(File.expand_path(File.dirname(__FILE__)))
```
Again, not sure why we need to chdir

er0ck replied:
in this case we might have to `chdir` because yard needs to work on code in a relative path.. and according to the `yard graph` manpage you can't specify a path to analyze.

er0ck commented:
@samwoods1 updated

RandellP commented on Rakefile:
```diff
@@ -1,29 +1,124 @@
+        puts "we made you a class diagram: #{DOCS_DIR}/arch_graph.png"
+      end
+    else
+      puts "ERROR: you don't have dot/graphviz; punting"
```
Is it really an error if it doesn't cause a failure.  Should it be a warning?

er0ck replied:
it fails to produce the image.  because it can't convert the dot format to png.
i could put that in stderr, if that helps, and exit 1

RandellP commented on Rakefile:
```diff
@@ -1,29 +1,124 @@
+    require "rspec/core/rake_task"
+    RSpec::Core::RakeTask.new(:spec)
+    desc "Run unit tests"
+    RSpec::Core::RakeTask.new do |t|
```
What does this second new do?  What task does it create?

er0ck replied:
actually, i think the first one is a duplication bug.  or really does nothing and is overwritten by the second which has some other config args.  i think i stripped this out in the gem_of version

RandellP commented on Rakefile:
```diff
@@ -1,29 +1,123 @@
+        puts "we made you a class diagram: #{DOCS_DIR}/arch_graph.png"
+      end
+    else
+      STDERR.puts "ERROR: you don't have dot/graphviz; punting"
```
To me the "punting" implied it was okay for it to fail, and things shouldn't stop.    The "ERROR" conflicted with that.  For me, I don't think generating the class diagram is really critical.  I don't think it gets checked in or anything.  So I would have gone the warning route.  But up to you.

RandellP commented on Rakefile:
```diff
@@ -1,29 +1,123 @@
+  begin
+    require "rspec/core/rake_task"
+    desc "Run unit tests"
+    RSpec::Core::RakeTask.new do |t|
```
Wouldn't hurt to comment on what the task name is here... or make it explicit in another way...

Commit:
(maint) turn off flog linting in CI, for now

* add more gems installed for the other rake tasks

Commit:
(maint) pin back some gems for use on the distelli hosts

Commit:
(maint) rescue missing rubycritic rake tasks in old ruby versions

* it's difficult to restrict bundle installed gems based upon ruby versions
* ensure the rakefile runs/works, but don't create a rubycritic task
* the rubycritic of yore also is super buggy, creates a bunch of cruft
in the path being evaluated, etc

Commit:
(maint) *sigh* clobber bundler over the head

Commit:
(maint) remove bundle cleanse to container build rake task

* this is handled in the Dockerfile
* ensure we still have clean bundle when running stuff in the container
  ** the update_and_start task copies all the files in again (including
  any Gemfile.lock)

Commit:
(maint) move distelli runs back to use rvm, local unit tests

* the rakefile uses gems' provided rake tasks
  * so we need the bundle anyway
  * the bundle is difficult to manage in the old ruby version on the
build host

Commit:
(maint) bump up code coverage

* add a comment on SLV-116 to fix tech-debt in `RototillerParam`
* add some tests around `#parent_name=`
* add some docs around `RototillerParam`'s methods

er0ck commented:
@samwoods1 @RandellP this is ready to go
